interfaces/builtin/pwm.go

```diff
@@ -0,0 +1,144 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/systemd"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+)
+
+// https://www.kernel.org/doc/Documentation/pwm.txt
+const pwmSummary = `allows access to specifc PWM channel`
+
+const pwmBaseDeclarationSlots = `
+  pwm:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
+    deny-auto-connection: true
+`
+
+var pwmSysfsPwmChipBase = "/sys/class/pwm/pwmchip%d"
+
+// pwmInterface type
+type pwmInterface struct {
+	commonInterface
+}
+
+// BeforePrepareSlot checks the slot definition is valid
+func (iface *pwmInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	// must have a PWM chanel number
+	number, ok := slot.Attrs["number"]
+	if !ok {
+		return fmt.Errorf("pwm slot must have a number attribute")
+	}
+
+	// valid values of number
+	if _, ok := number.(int64); !ok {
+		return fmt.Errorf("pwm slot number attribute must be an int")
+	}
+
+	// must have a PWM chip number
+	chipNum, ok := slot.Attrs["chip-number"]
+	if !ok {
+		return fmt.Errorf("pwm slot must have a chip-number attribute")
+	}
+
+	// valid values of chip number
+	if _, ok := chipNum.(int64); !ok {
+		return fmt.Errorf("pwm slot chip-number attribute must be an int")
+	}
+
+	// slot is good
+	return nil
+}
+
+func (iface *pwmInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var chipNum int64
+	if err := slot.Attr("chip-number", &chipNum); err != nil {
+		return err
+	}
+
+	var number int64
+	if err := slot.Attr("number", &number); err != nil {
+		return err
+	}
+	path := fmt.Sprintf(pwmSysfsPwmChipBase, chipNum)
+	// Entries in /sys/class/pwm for PWM chips are just symlinks
+	// to their correct device part in the sysfs tree. Given AppArmor
+	// requires symlinks to be dereferenced, evaluate the PWM
+	// path and add the correct absolute path to the AppArmor snippet.
+	dereferencedPath, err := evalSymlinks(path)
+	if err != nil && os.IsNotExist(err) {
+		// If the specific pwm is not available there is no point
+		// exporting it, we should also not fail because this
+		// will block snapd updates (LP: 1866424)
+		logger.Noticef("cannot find not existing pwm chipbase %s", path)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	spec.AddSnippet(fmt.Sprintf("%s/pwm%d/* rwk,", dereferencedPath, number))
```

Comment:
I wonder what your thoughts would be on having explicit entries for the various files under `pwm%d` with the appropriate permissions - ie `period`, `duty_cycle`, `polarity` and `enable` rather than just `*` - otherwise if in a future kernel update there are new files added here that grant extra permissions this will be implicitly granted access as well?

Reply:
fwiw gpio has also /* there

Reply:
Yes, I saw that precedent, but I thought perhaps it might be better to take a more defensive approach - and so I wondered what others opinions on this would be? I prefer to be more specific with these interfaces but if the general approach is to be less specific and more flexible (ie `*` in this case) that is fine, I just want it to be a conscious choice and to make sure we take a consistent approach when adding new interfaces like this in the future.

Reply:
Personally, I think for an interface like this which already grants write access to low level devices and doesn't auto-connect, I think it's okay to be more permissive with `*`, but if it were a read-only interface or something that auto-connected I think we should be more specific, but I will defer to @pedronis on the general direction we want to take going forward on these interfaces

Reply:
that access at the end is scoped down to a pin/channel so unless the kernel makes very unorthodox choices the relevant surface should be contained

Reply:
I agree with Ian here and think this is okay. I will merge given that it seems this is not super controversial, then we can always tweak/improve before this lands on master.